### ISSUE TITLE
Two closures as function call arguments with `AssertUnwindSafe`

### ISSUE THREAD

**Issue by Enyium** [rust-lang/rustfmt#5678]
For situations without `AssertUnwindSafe`, this was already solved in #1713.

But when you need to mark a closure with [`AssertUnwindSafe`](https://doc.rust-lang.org/std/panic/struct.AssertUnwindSafe.html), the formatting is as bad as before above issue was solved:

```rust
Plugin::catch_unwind_panic(AssertUnwindSafe(|| plugin_fn(args, env)), || {
    format!("`{fn_name}()'")
});
```

I guess in my case it should be formatted like this:

```rust
Plugin::catch_unwind_panic(
    AssertUnwindSafe(|| plugin_fn(args, env)),
    || format!("`{fn_name}()'"),
);
```

**Comment by ytmimi** [comment]
Thanks for reaching out, but this is different from #1713, which focused on calls with multiple closures.

Your example has an associated function call `Plugin::catch_unwind_panic` with two arguments. The first of which is also a function call `AssertUnwindSafe(|| plugin_fn(args, env))` and the second is a closure ``|| { format!("`{fn_name}()'")}``. At the AST level these are all just expressions and what you've got here generalizes to `expr(expr, expr)`.

Check out the last example, listed in the [`Combinable expressions`](https://github.com/rust-lang/fmt-rfcs/blob/master/guide/expressions.md#combinable-expressions) section of the [Rust Style Guide](https://github.com/rust-lang/fmt-rfcs/blob/master/guide/guide.md), and you'll see a prescription for how to handle closures with block bodies when they're the last argument to a function. I understand that the formatting here isn't ideal, but the behavior is correct.

If you'd like you can use the `#[rustfmt::skip]` attribute after hand formatting the call to achieve your desired formatting and to prevent rustfmt from modifying it going forward.

Alternatively you may choose to refactor the code to not define the closure inline, which might be a little more readable.

```rust
let first_closure = || plugin_fn(args, env);
let trailing_closure = || format!("`{fn_name}()'");
Plugin::catch_unwind_panic(AssertUnwindSafe(first_closure), trailing_closure);
```

**Comment by Enyium** [comment]
> The first of which is also a function call `AssertUnwindSafe(|| plugin_fn(args, env))`

[`std::panic::AssertUnwindSafe`](https://doc.rust-lang.org/std/panic/struct.AssertUnwindSafe.html) is a struct, not a function (see first example behind link). I would argue `AssertUnwindSafe` can be treated as special, which is not like *any* struct instantiation, because it is used as a marker. I guess there's compiler magic involved to make it work.

**Comment by ytmimi** [comment]
> [std::panic::AssertUnwindSafe](https://doc.rust-lang.org/std/panic/struct.AssertUnwindSafe.html) is a struct, not a function.

Correct. What I was trying to convey was that from the perspective of the AST we get back from the rustc_parser `AssertUnwindSafe(|| plugin_fn(args, env))` will be parsed as a function call.

>  I would argue AssertUnwindSafe can be treated as special

I'm not sure If I agree with that. I think context is really important, and in the context of your code snippet above the `AssertUnwindSafe` formatting is hard to parse for humans.

In a different context the formatting might be much friendlier on the eyes. Here's an example copied from the rustfmt codebase where `AssertUnwindSafe` is the only argument to a function call.

```rust
        let result = catch_unwind(AssertUnwindSafe(|| {
            rewrite_macro_inner(
                mac,
                extra_ident,
                context,
                shape,
                position,
                guard.is_nested(),
            )
        }));
```

Also, here's the formatting you'd get if we switched the  order of the `AssertUnwindeSafe` and the closure in your example snippet above:

```rust
    Plugin::catch_unwind_panic(
        || format!("`{fn_name}()'"),
        AssertUnwindSafe(|| plugin_fn(args, env)),
    )
```

Which is more or less what you're looking for.

**Comment by calebcartwright** [comment]
I'm going to also weigh in here and say unequivocally that we're not going to make any changes.

As Yacin has already noted, there's only two args, and one only one of those (the last one) is a closure, so it's formatted exactly how the Style Guide prescribes and that's really where things end for us on the rustfmt team.

If you'd like to propose a new rule or modification to the existing rule, then the correct place to do that would be https://github.com/rust-lang/fmt-rfcs. It's certainly possible that request could be taken into consideration for a future edition of the Style Guide, but it's not a change that could ever be made to the current edition due to stability guarantees.

FWIW I wouldn't be too terribly optimistic about the likelihood of your request making it into a future edition either because it could quickly turn into a slippery slope. However, it never hurts to ask/suggest!

Finally, I'd also point out that you could also achieve the formatting you want via the inclusion of a line comment in various places within the `catch_unwind_panic` call, e.g.:

```rust
fn foo() {
    Plugin::catch_unwind_panic(
        AssertUnwindSafe(|| plugin_fn(args, env)),
        // bar
        || format!("`{fn_name}()'"),
    );
}

fn foo() {
    Plugin::catch_unwind_panic(
        // foo
        AssertUnwindSafe(|| plugin_fn(args, env)),
        || format!("`{fn_name}()'"),
    );
}
```

**Comment by Enyium** [comment]
I wouldn't get into RFCs for this. It's just one location in my code.

As a compromise, I now use an empty line comment. I liked the other options even less. Thanks for pointing that out.
